assets/js/blocks/product-search/block.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import classnames from 'classnames';
+import { Component } from '@wordpress/element';
+import { IconButton } from '@wordpress/components';
+import PropTypes from 'prop-types';
+import { withInstanceId, compose } from '@wordpress/compose';
+import { RichText } from '@wordpress/editor';
+
+/**
+ * Internal dependencies
+ */
+import './editor.scss';
+import './style.scss';
+
+/**
+ * Component displaying a product search form.
+ */
+class ProductSearchBlock extends Component {
+	render() {
+		if ( this.props.isPreview ) {
+			return this.renderEdit();
+		}
+
+		return this.renderView();
+	}
+
+	renderView() {
+		const { attributes: { label, placeholder, formId, className, hasLabel } } = this.props;
+		const home = wc_product_block_data.homeUrl;
+		const classes = classnames(
+			'wc-block-product-search',
+			className,
+		);
+		const data = {};
+
+		if ( hasLabel ) {
+			data[ 'data-has-label' ] = true;
+		}
+
+		return (
+			<form className={ classes } { ...data } role="search" method="get" action={ home } data-form-id={ formId }>
+				<label
+					htmlFor={ formId }
+					className={ hasLabel ? 'wc-block-product-search__label' : 'wc-block-product-search__label screen-reader-text' }
+				>
+					{ label }
+				</label>
+				<div className="wc-block-product-search__fields">
+					<input
+						type="search"
+						id={ formId }
+						className="wc-block-product-search__field"
+						placeholder={ placeholder }
+						name="s"
+					/>
+					<input type="hidden" name="post_type" value="product" />
+					<IconButton
```

Comment:
Considering we are saving this to the markup, I think we shouldn't use a WordPress component but a plain HTML button. Similarly to what you did for the _Product Categories List_ block.